assignment2.rb

```diff
@@ -0,0 +1,132 @@
+#Blocks are used for inline code within method invocations, 
+#procs are general-purpose saved blocks,
+# and lambdas are special type of procs with stricter argument handling.
+
+#Use Blocks, Procs and Lambda each to write a program which takes array of integers as an input and finds the count of even numbers in the array
+
+
+#using Blocks
+def count_even_numbers(array)
+    count = 0
+    array.each do |num|
+      count += 1 if num.even?
+    end
+    count
+  end
+  
+ numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+  puts "Count of even numbers: #{count_even_numbers(numbers)}"
+
+  #using Procs
+  count_even = Proc.new { |num| num.even? }
+def count_numbers(array, condition)
+  array.count(&condition)
+end
+
+
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+puts "Count of even numbers: #{count_numbers(numbers, count_even)}"
+
+#using lambda
+
+count_even = lambda { |num| num.even? }
+def count_numbers(array, condition)
+  array.count(&condition)
+end
+
+
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+puts "Count of even numbers: #{count_numbers(numbers, count_even)}"
+
+
+#####################################################################################################################################################
+
+require 'csv'
+
+class Employee
+  attr_accessor :name, :age, :department, :division, :designation
+
+  def initialize(name, age, department, division, designation)
+    @name = name
+    @age = age
+    @department = department
+    @division = division
+    @designation = designation
+  end
+end
+
+class Hr
+    attr_reader :employees
+  
+    def initialize
+      @employees = []
+    end
+  
+    def add_employee(name, age, department, division, designation)
+      employee = Employee.new(name, age, department, division, designation)
+      @employees << employee
+      save_to_csv
+    end
+  
+    def get_all_employees
+      @employees
+    end
+  
+    private
+  
+    def save_to_csv
+      CSV.open('employees.csv', 'w') do |csv|
+        csv << ['Name', 'Age', 'Department', 'Division', 'Designation']
+        @employees.each do |employee|
+            puts employee
+          csv << [employee.name, employee.age, employee.department, employee.division, employee.designation]
+        end
+      end
+    end
+  end
+
+  class Department
+    attr_reader :name, :department_head, :employees
+  
+    def initialize(name, department_head)
+      @name = name
+      @department_head = department_head
+      @employees = []
+    end
+  
+    def add_employee(employee)
+      @employees << employee
+    end
+  
+    def get_department_employees
+      @employees
+    end
+  end
+
+  class Project
+    attr_accessor :type, :revenue, :profit
+  
+    def initialize(type, revenue, profit)
+      @type = type
+      @revenue = revenue
+      @profit = profit
+    end
+  end
+
+  hr = Hr.new
```

Comment:
I don't see `Company` class at all in this code. Please check each and every point of the assignment again.

Reply:
Added company class and worked on the points
Only the HR can add/edit details of the employee.

Only the HR can get the details of all the employees. 

The department heads can get the details of those employees in his department.

Projects in the company can either be a Service or Product

Each service offered by the company has 1 external client 

Each product developed by the company has several clients. 

The company keeps track of the yearly revenue and profit for the projects. The profit for service & product is calculated differently.